### PR TITLE
Clean up QK and file and tensor types

### PR DESCRIPTION
**Pull request by sw**
(edit: the python changes will clash with #545)

This PR has several goals:
* reduce the number of integer literals strewn across the code base
* better ensure array and enum consistency
* clear up possible confusion about file and tensor types
* prepare the way for new quantization formats with QK != 32, as discussed in #456
* deduplicate python definitions

For the python scripts, I introduce a new file `ggml.py` at the top level, which contains definitions of the file and tensor types equivalent to those in ggml.h. I have formatted that with `black`, in case #611 returns from the dead.

The changes to the python files on one hand and C/C++ on the other are technically independent, but the discussion will overlap, so I'm keeping this in one draft. I will split it later if it makes sense.

I have tested conversion from pth to ggml with identical outputs, but I have not tested the other conversion scripts.

Open questions:
* should `enum e_ftype` be moved to `llama.h` (with sensible renaming)? This would allow us to eliminate the hard-coded 2,3 in the usage string of `quantize.cpp`, and maybe be useful elsewhere.
* how is file type 4 used? ~~I could not find any other parts of the code that would use this.~~ I see now, it's for GPTQ models. Should I add `FTYPE_GPTQ`?. https://github.com/ggerganov/llama.cpp/blob/3525899277d2e2bdc8ec3f0e6e40c47251608700/llama.cpp#L515
* Is `GGML_FILE` ok or should it be `LLAMA_FILE`? ggml.c doesn't deal with that type.

Some more comments below, looking for your thoughts on this...

### CODE REVIEW

**Comment by prusnak** [comment]
I think it would be worthwhile to separate the Python changes into a separate PR which can be reviewed and merged sooner than the more complex changes in C.

File: convert-gpt4all-to-ggml.py [hunk 1]
```diff
@@ -32,6 +33,7 @@ def write_header(f_out, header):
 
     if magic != 0x67676d6c:
```
Comment (j-f1) [inline convert-gpt4all-to-ggml.py:34]: Can you make the magic a constant too?

Comment (prusnak) [reply convert-gpt4all-to-ggml.py:34]: We need old magic constants anyway to detect older models.

File: llama.cpp [hunk 2]
```diff
@@ -100,7 +109,7 @@ struct llama_hparams {
     int32_t n_head  = 32;
     int32_t n_layer = 32;
     int32_t n_rot   = 64;
-    int32_t f16     = 1;
+    int32_t f16     = FTYPE_F16;
```
[the remark this answers is not in the document]
Comment (ggerganov) [reply llama.cpp:112]: `f16` is badly named since the times I was only considering F16 and F32 and no other data types

**Review by ggerganov — COMMENTED** [review]
These refactors and code maintenance changes are very helpful

- `enum e_ftype` can be moved to `llama.h` and be called `enum llama_ftype`
- I think `GGML_FILE` -> `LLAMA_FTYPE`

**Pull request closed without merging**